Sources/SwiftLexicalLookup/SequentialScopeSyntax.swift

```diff
@@ -0,0 +1,108 @@
+  ///
+  /// Example:
+  /// ```swift
+  /// func foo() {
+  ///   let a = 1
+  ///   guard let a = x else { return }
+  ///   let a = a // <-- 1
+  ///   guard let a = y else { return }
+  ///   a // <-- 2
+  /// }
+  /// ```
+  /// For the first `a` reference, sequential lookup returns
+  /// two results: from `guard` scope and from code block scope
+  /// in this exact order. For the second `a` reference,
+  /// sequential lookup yields results from four scopes starting
+  /// from the bottom: `guard`, code block, `guard` and
+  /// code block scope in this exact order.
+  func sequentialLookup(
+    in codeBlockItems: some Collection<CodeBlockItemSyntax>,
+    for identifier: Identifier?,
+    at origin: SyntaxProtocol,
+    with config: LookupConfig,
+    state: LookupState,
+    createResultsForThisScopeWith getResults: ([LookupName]) -> (LookupResult)
+  ) -> [LookupResult] {
+    var result: [LookupResult] = []
+    var currentChunk: [LookupName] = []
+    var itemsWithoutNamedDecl: [CodeBlockItemSyntax] = []
+
+    for codeBlockItem in codeBlockItems {
+      if Syntax(codeBlockItem.item).isProtocol(NamedDeclSyntax.self) {
+        currentChunk += LookupName.getNames(
+          from: codeBlockItem.item,
+          accessibleAfter: codeBlockItem.endPosition
+        ).filter { introducedName in
+          checkName(identifier, refersTo: introducedName, at: origin)
+        }
+      } else {
+        itemsWithoutNamedDecl.append(codeBlockItem)
+      }
+    }
+
+    for codeBlockItem in itemsWithoutNamedDecl {
+      guard codeBlockItem.position < origin.position else { break }
+
+      if let introducingToParentScope = Syntax(codeBlockItem.item).asProtocol(SyntaxProtocol.self)
+        as? IntroducingToSequentialParentScopeSyntax
+      {
+        // Get results from encountered scope.
+        let introducedResults = introducingToParentScope.introducesToSequentialParent(
+          for: identifier,
+          at: origin,
+          with: config,
+          state: state
+        )
+
+        // Skip, if no results were found.
+        guard !introducedResults.isEmpty else { continue }
+
+        // If there are some names collected, create a new result for this scope.
+        if !currentChunk.isEmpty {
+          result.append(getResults(currentChunk))
+          currentChunk = []
+        }
+
+        result += introducedResults
+      } else {
+        // Extract new names from encountered node.
+        currentChunk += LookupName.getNames(
+          from: codeBlockItem.item,
+          accessibleAfter: codeBlockItem.endPosition
+        ).filter { introducedName in
+          checkName(identifier, refersTo: introducedName, at: origin)
+        }
+      }
+    }
+
+    // If there are some names collected, create a new result for this scope.
+    if !currentChunk.isEmpty {
+      result.append(getResults(currentChunk))
+    }
+
+    return result.reversed() + lookupInParent(for: identifier, at: origin, with: config, state: state)
```

Comment:
Instead of reversing the result here (which requires a copy of the array), have you considered changing the semantics of the result to say a later element overrides earlier elements? I think I would also find that more intuitive because it matches the source order.

Reply:
As we already represent results from inner-most to outer-most, I think it makes most sense to reverse it. Like in [this example](https://github.com/swiftlang/swift-syntax/blob/c83b3c6f3454a05b0a22ce26f12de594de19d68b/Tests/SwiftLexicalLookupTest/NameLookupTests.swift#L803-L832):
```Swift
func foo() {
  let a = 1 // <-- Code block scope
  let b = 2

  guard let b = a else { return }

  let a = 3 // <-- Code block scope
  let b = 4

  guard let a = b else { return } // <-- Guard scope

  print(a, b) // <-- lookup for a
}
```
I think the order of results should be: `guard` scope and exactly one code block scope result (as the first `guard` is not relevant for `a` lookup).

The order of `LookupName`s within `LookupResult`s is the way you described actually. In the result from code block scope the order is exactly: `let a = 1` and then `let a = 3`.